framework/src/org/checkerframework/framework/type/SupertypeFinder.java

```diff
@@ -161,10 +161,17 @@
                 }
             }
 
-            for (int i = 0; i < type.getTypeArguments().size(); ++i) {
-                AnnotatedTypeMirror typArg = type.getTypeArguments().get(i);
-                TypeParameterElement ele = typeElement.getTypeParameters().get(i);
-                mapping.put(ele, typArg);
+            AnnotatedDeclaredType enclosing = type;
+            while (enclosing != null) {
+                TypeElement enclosingTypeElement =
+                        (TypeElement) enclosing.getUnderlyingType().asElement();
+                for (int i = 0; i < enclosing.getTypeArguments().size(); ++i) {
+                    AnnotatedTypeMirror typArg = enclosing.getTypeArguments().get(i);
+                    TypeParameterElement ele = enclosingTypeElement.getTypeParameters().get(i);
```

Comment:
Note that `getTypeParameters` allocates a fresh list on every call. I would suggest putting both the type arguments and the type parameters into locals outside the loop.